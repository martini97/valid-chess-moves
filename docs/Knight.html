<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Class: Knight</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Class: Knight</h1>

    




<section>

<header>
    
        <h2><span class="attribs"><span class="type-signature"></span></span>Knight<span class="signature">(algebraicPosition)</span><span class="type-signature"></span></h2>
        
            <div class="class-description">Class representing a Knight piece.</div>
        
    
</header>

<article>
    <div class="container-overview">
    
        

    
    <h2>Constructor</h2>
    

    
    <h4 class="name" id="Knight"><span class="type-signature"></span>new Knight<span class="signature">(algebraicPosition)</span><span class="type-signature"></span></h4>
    

    











    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>algebraicPosition</code></td>
            

            <td class="type">
            
                
<span class="param-type"><a href="global.html#AlgebraicNotation">AlgebraicNotation</a></span>


            
            </td>

            

            

            <td class="description last">The algebraic position of the Knight.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="knight.js.html">knight.js</a>, <a href="knight.js.html#line8">line 8</a>
    </li></ul></dd>
    

    

    

    
</dl>


















    
    </div>

    
        <h3 class="subsection-title">Extends</h3>

        


    <ul>
        <li><a href="Piece.html">Piece</a></li>
    </ul>


    

    

    

    

    

    

    
        <h3 class="subsection-title">Members</h3>

        
            
<h4 class="name" id="algebraicPosition"><span class="type-signature"></span>algebraicPosition<span class="type-signature"> :<a href="global.html#AlgebraicNotation">AlgebraicNotation</a></span></h4>




<div class="description">
    The algebraic position of the piece.
</div>



    <h5>Type:</h5>
    <ul>
        <li>
            
<span class="param-type"><a href="global.html#AlgebraicNotation">AlgebraicNotation</a></span>


        </li>
    </ul>





<dl class="details">

    

    

    
    <dt class="inherited-from">Inherited From:</dt>
    <dd class="inherited-from"><ul class="dummy"><li>
        <a href="Piece.html#algebraicPosition">Piece#algebraicPosition</a>
    </li></ul></dd>
    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="piece.js.html">piece.js</a>, <a href="piece.js.html#line14">line 14</a>
    </li></ul></dd>
    

    

    

    
</dl>






        
            
<h4 class="name" id="coordinates"><span class="type-signature"></span>coordinates<span class="type-signature"> :<a href="global.html#Coordinates">Coordinates</a></span></h4>




<div class="description">
    The coordinates of the piece.
</div>



    <h5>Type:</h5>
    <ul>
        <li>
            
<span class="param-type"><a href="global.html#Coordinates">Coordinates</a></span>


        </li>
    </ul>





<dl class="details">

    

    

    
    <dt class="inherited-from">Inherited From:</dt>
    <dd class="inherited-from"><ul class="dummy"><li>
        <a href="Piece.html#coordinates">Piece#coordinates</a>
    </li></ul></dd>
    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="piece.js.html">piece.js</a>, <a href="piece.js.html#line20">line 20</a>
    </li></ul></dd>
    

    

    

    
</dl>






        
            
<h4 class="name" id="rules"><span class="type-signature"></span>rules<span class="type-signature"> :Array.&lt;<a href="global.html#Coordinates">Coordinates</a>></span></h4>


<p class="summary">The rules for the piece movement.</p>



<div class="description">
    This field defines the overall rules for the movement of a piece,
override it on the specific piece class with the piece rule.
</div>



    <h5>Type:</h5>
    <ul>
        <li>
            
<span class="param-type">Array.&lt;<a href="global.html#Coordinates">Coordinates</a>></span>


        </li>
    </ul>





<dl class="details">

    

    

    

    
    <dt class="tag-overrides">Overrides:</dt>
    <dd class="tag-overrides"><ul class="dummy"><li>
        <a href="Piece.html#rules">Piece#rules</a>
    </li></ul></dd>
    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="piece.js.html">piece.js</a>, <a href="piece.js.html#line29">line 29</a>
    </li></ul></dd>
    

    

    

    
</dl>






        
    

    
        <h3 class="subsection-title">Methods</h3>

        
            

    

    
    <h4 class="name" id="possiblePositionsInNMoves"><span class="type-signature"></span>possiblePositionsInNMoves<span class="signature">(movesInfo)</span><span class="type-signature"> &rarr; {Array.&lt;<a href="global.html#AlgebraicNotation">AlgebraicNotation</a>>}</span></h4>
    

    



<div class="description">
    All the possible (legal) movements of a piece after n moves.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>movesInfo</code></td>
            

            <td class="type">
            
                
<span class="param-type">Object</span>


            
            </td>

            

            

            <td class="description last">Object containing the params.
                <h6>Properties</h6>
                

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        
        <th>Attributes</th>
        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>startAt</code></td>
            

            <td class="type">
            
                
<span class="param-type"><a href="global.html#Coordinates">Coordinates</a></span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    this.coordinates
                
                </td>
            

            <td class="description last">Starting position of the
piece. Defaults to the position of the current piece.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>n</code></td>
            

            <td class="type">
            
                
<span class="param-type">Integer</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    1
                
                </td>
            

            <td class="description last">Number of moves, where 0 < n < 20.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>positions</code></td>
            

            <td class="type">
            
                
<span class="param-type">Array.&lt;<a href="global.html#Coordinates">Coordinates</a>></span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    []
                
                </td>
            

            <td class="description last">Array of positions. Used for recursion.</td>
        </tr>

    
    </tbody>
</table>

            </td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    
    <dt class="inherited-from">Inherited From:</dt>
    <dd class="inherited-from"><ul class="dummy"><li>
        <a href="Piece.html#possiblePositionsInNMoves">Piece#possiblePositionsInNMoves</a>
    </li></ul></dd>
    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="piece.js.html">piece.js</a>, <a href="piece.js.html#line104">line 104</a>
    </li></ul></dd>
    

    

    

    
</dl>











<h5>Throws:</h5>

        

<dl>
    <dt>
        <div class="param-desc">
        Throws a BadRequestError if the given n is invalid.
        </div>
    </dt>
    <dd></dd>
    <dt>
        <dl>
            <dt>
                Type
            </dt>
            <dd>
                
<span class="param-type">BadRequestError</span>


            </dd>
        </dl>
    </dt>
    <dd></dd>
</dl>


    


<h5>Returns:</h5>

        
<div class="param-desc">
    Every possible coordinates that the piece can reach,
it includes coordinates out of the board.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">Array.&lt;<a href="global.html#AlgebraicNotation">AlgebraicNotation</a>></span>


    </dd>
</dl>

    





        
    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="Knight.html">Knight</a></li><li><a href="Piece.html">Piece</a></li></ul><h3><a href="global.html">Global</a></h3>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.5.5</a> on Sun Jul 15 2018 16:48:18 GMT-0300 (Brasilia Standard Time)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>