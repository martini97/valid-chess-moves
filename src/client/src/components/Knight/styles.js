export default theme => ({
  knight: {
    fontSize: '10vmin',
    position: 'absolute',
    top: '50%',
    textAlign: 'center',
    transform: 'translateY(-50%)',
  },
});
