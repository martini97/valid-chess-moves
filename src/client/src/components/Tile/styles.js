import { grey, blue } from '@material-ui/core/colors';

export default theme => ({
  tile: {
    position: 'relative',
    height: '100%',
    paddingRight: `${100/8}%`,
    cursor: 'pointer',
  },
  black: {
    backgroundColor: grey[900],
  },
  white: {
    backgroundColor: grey[400],
  },
  blackPossible: {
    backgroundColor: blue[900],
  },
  whitePossible: {
    backgroundColor: blue[400],
  },
});
