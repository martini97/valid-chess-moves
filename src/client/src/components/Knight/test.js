import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import ReactTestUtils from 'react-dom/test-utils';
import { render, mount } from 'enzyme';

import Knight from '.';

it('renders without crashing', () => {
  const div = document.createElement('div');

  ReactDOM.render(<Knight />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('snapshot', () => {
  const component = renderer.create(<Knight />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
