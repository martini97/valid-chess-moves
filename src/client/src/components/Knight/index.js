import React from 'react';
import Proptypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';

import styles from './styles';

const Knight = ({ classes }) => (
  <span role="img" aria-label="horse" className={classes.knight}>
    🐴
  </span>
);

Knight.propTypes = {
  classes: Proptypes.object.isRequired,
};

export default withStyles(styles)(Knight);
