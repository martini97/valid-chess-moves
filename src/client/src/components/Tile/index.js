import React from 'react';
import Proptypes from 'prop-types';
import classnames from 'classnames';
import { Grid } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';

import Knight from '../Knight';
import styles from './styles';

function tileColor(column, row, possible) {
  const color = (column.charCodeAt(0) - 96 + row) % 2 ? 'black' : 'white';
  return possible ? `${color}Possible` : color
}

const Tile = ({ classes, row, column, selected, possible, disabled, handleClick }) => {
  const tileClass = classnames(classes.tile, classes[tileColor(column, row, possible)]);
  return (
    <Grid item className={tileClass} onClick={disabled ? () => null : handleClick}>
      { selected && (<Knight />) }
    </Grid>
  )
};

Tile.propTypes = {
  classes: Proptypes.object.isRequired,
  row: Proptypes.number.isRequired,
  column: Proptypes.string.isRequired,
  handleClick: Proptypes.func.isRequired,
  selected: Proptypes.bool,
  possible: Proptypes.bool,
  disabled: Proptypes.bool,
};

Tile.defaultProps = {
  selected: false,
  possible: false,
  disabled: false,
};

export default withStyles(styles)(Tile);
