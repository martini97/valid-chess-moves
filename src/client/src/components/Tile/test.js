import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import ReactTestUtils from 'react-dom/test-utils';
import { render, mount } from 'enzyme';

import Knight from '../Knight';

import Tile from '.';

it('renders without crashing', () => {
  const div = document.createElement('div');

  const props = {
    row: 1,
    column: 'a',
    handleClick: () => null,
  };

  ReactDOM.render(<Tile {...props} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('snapshot', () => {
  const props = {
    row: 1,
    column: 'a',
    handleClick: () => null,
  };

  let component = renderer.create(<Tile {...props} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  props.selected = true;
  component = renderer.create(<Tile {...props} />);
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  props.possible = true;
  component = renderer.create(<Tile {...props} />);
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  props.selected = true;
  component = renderer.create(<Tile {...props} />);
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should not contain Knight when selected', () => {
  const props = {
    row: 1,
    column: 'a',
    handleClick: () => null,
    selected: false,
  };
  const wrapper = mount(<Tile {...props} />);
  expect(wrapper.contains(Knight)).toEqual(false);
});

it('should contain Knight when selected', () => {
  const props = {
    row: 1,
    column: 'a',
    handleClick: () => null,
    selected: true,
  };
  const wrapper = mount(<Tile {...props} />);
  expect(wrapper.contains(Knight)).toEqual(true);
});

it('should call function on click', () => {
  const props = {
    row: 1,
    column: 'a',
    handleClick: jest.fn(),
    selected: true,
  };
  const wrapper = mount(<Tile {...props} />);

  wrapper.simulate('click');
  expect(props.handleClick.mock.calls.length).toBe(1);
});

it('should not call function on click when disabled', () => {
  const props = {
    row: 1,
    column: 'a',
    handleClick: jest.fn(),
    disabled: true,
  };
  const wrapper = mount(<Tile {...props} />);

  wrapper.simulate('click');
  expect(props.handleClick.mock.calls.length).toBe(0);
});
