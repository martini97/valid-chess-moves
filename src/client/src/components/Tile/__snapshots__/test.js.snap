// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  className="MuiGrid-item-108 Tile-tile-102 Tile-white-104"
  onClick={[Function]}
/>
`;

exports[`snapshot 2`] = `
<div
  className="MuiGrid-item-108 Tile-tile-102 Tile-white-104"
  onClick={[Function]}
>
  <span
    aria-label="horse"
    className="Knight-knight-203"
    role="img"
  >
    🐴
  </span>
</div>
`;

exports[`snapshot 3`] = `
<div
  className="MuiGrid-item-108 Tile-tile-102 Tile-whitePossible-106"
  onClick={[Function]}
>
  <span
    aria-label="horse"
    className="Knight-knight-203"
    role="img"
  >
    🐴
  </span>
</div>
`;

exports[`snapshot 4`] = `
<div
  className="MuiGrid-item-108 Tile-tile-102 Tile-whitePossible-106"
  onClick={[Function]}
>
  <span
    aria-label="horse"
    className="Knight-knight-203"
    role="img"
  >
    🐴
  </span>
</div>
`;
